doc/getdns_general.3.in
.\" The "BSD-New" License
.\" 
.\" Copyright (c) 2013, NLnet Labs, Verisign, Inc.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\" * Redistributions of source code must retain the above copyright
.\"   notice, this list of conditions and the following disclaimer.
.\" * Redistributions in binary form must reproduce the above copyright
.\"   notice, this list of conditions and the following disclaimer in the
.\"   documentation and/or other materials provided with the distribution.
.\" * Neither the names of the copyright holders nor the
.\"   names of its contributors may be used to endorse or promote products
.\"   derived from this software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED. IN NO EVENT SHALL Verisign, Inc. BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 

.TH getdns_general 3 "@date@" "getdns @version@" getdns
.SH NAME
.B getdns_general, 
.B getdns_general_sync
-- do a getdns DNS lookup

.SH LIBRARY
DNS Resolver library (libgetdns, -lgetdns)

.SH SYNOPSIS
#include <getdns.h>

getdns_return_t 
.br
.B getdns_general
(getdns_context *context,
.RS 3
const char *name,
.br
uint16_t request_type
.br
getdns_dict *extensions,
.br
void *userarg,
.br
getdns_transaction_t *transaction_id,
.br
getdns_callback_t callbackfn)
.RE

getdns_return_t 
.br
.B getdns_general_sync
(getdns_context *context,
.RS 3
const char *name,
.br
uint16_t request_type
.br
getdns_dict *extensions,
.br
getdns_dict **response)
.RE

.SH DESCRIPTION

.LP
The getdns_general(3) and getdns_general_sync functions provide public entry
points into the getdns API library to retrieve any valid responses to a query
from the DNS (note that other namespaces in the context are not used).   Most
typical use cases for applications are probably satisifed via calls to
getdns_address(3) which would replace getaddrinfo(3).

.HP 3
.I context
A pointer to the previsouly created DNS context that is to be used with this DNS request. see getdns_context (3)

.HP 3
.I name
The ASCII-based domain name looked up as a string. This can also be an
IPv4 or IPv6 address for request types that take addresses instead of domain
names, such as PTR. The values here follow the rules in section 2.1 of RFC 4343
to allow non-ASCII octets and special characters in labels.

.HP 3
.I request_type
Specifies the RRtype for the query; the RRtype numbers are listed in the IANA registry. For example, to get the NS records, request_type would be 2. The API also has defined macros for most of the RRtypes by name; the definition names all start with "GETDNS_RRTYPE_". For example, to get the NS records, you can also set the request_type to GETDNS_RRTYPE_NS.

.HP 3
.I extensions
extensions for this request, NULL if no extensions, see libgetnds (3) for a detailed description of extensions

.HP 3
.I userarg
returned to the callback function untouched, can be NULL

.HP 3
.I transaction_id
populated by the API and used to identify the callback (for example to getdns_cancel_callback), can be NULL, set to 0 if the function fails

.HP 3
.I callbackfn
non-NULL pointer to a callback function defined by the application, typically
used to process the response. Only the asynchronous signature accepts a
callback function, the synchronous signature does not include a callback.  See
libgetdns (3) for a more detailed discussion of callback functions.

.HP 3
.I response
A getdns_dict type is returned in response and always contains at least three names: replies_full (a list containing the DNS response as binary data), replies_tree (a list containing the parsed DNS response data) and status (an int).  The storage associated with this must be freed by a call to getdns_free_sync_request_memory (3).

.HP
.SH "RETURN VALUES"

Upon successful completion the functions return
.B GETDNS_RETURN_GOOD
, otherwise the following error values are returned:

.LP
.B GETDNS_RETURN_BAD_CONTEXT 
if the context pointer is invalid or the context has internal deficiencies
.LP
.B GETDNS_RETURN_BAD_DOMAIN_NAME
if the domain name passed to the function is invalid
.LP
.B GETDNS_RETURN_EXTENSION_MISFORMAT
if the data type specified in one or more of the extensions does not match the specifications
.LP
.B GETDNS_RETURN_GENERIC_ERROR
if some problem was encountered in the function not addressed by one of the more
specific return codes
.LP
.B GETDNS_RETURN_INVALID PARAMETER 
if one or more parameters has an invalid value
.LP
.B GETDNS_RETURN_MEMORY_ERROR
if unable to allocate the memory required
.LP
.B GETDNS_RETURN_NO_SUCH_EXTENSION
if one or more of the strings specified in the extensions are not valid

The values of status included in the response parameter are:

.LP
.B GETDNS_RESPSTATUS_GOOD
if at least one response was returned
.LP
.B GETDNS_RESPSTATUS_NO_NAME
if queries for the name yielded all negative responses
.LP
.B GETDNS_RESPSTATUS_ALL_TIMEOUT
if all queries for the name timed out
.LP
.B GETDNS_RESPSTATUS_NO_SECURE_ANSWERS
if only secure replies accepted (per context) and at least one response was received but no DNS responses were secure through DNSSEC

.LP
For a more detailed explanation of the response object see
.I libgetdns
(3)

.SH REQUEST TYPES

This is a list of the most common request types, a full list of request types in more detail is available at http://www.iana.org/assignments/dns-parameters/dns-parameters.xml

.RS 3
.TP 11
.B A
Host address
.TP 
.B AAAA
IPv6 address
.TP
.B CAA
Certificate Authority Authorization
.TP
.B CNAME
Canonical name for an alias
.TP
.B DLV
DNSSEC lookaside validation
.TP
.B DNAME
DNAME
.TP
.B DS
Delegation signer
.TP
.B HINFO
Host information
.TP
.B KEY
Security key
.TP
.B MINFO
Mailbox or mail list information
.TP
.B MX
Mail exchange
.TP
.B NS
Authoritative name server
.TP
.B NSEC
Next secure record
.TP
.B NSEC3
Next secure record (hashed)
.TP
.B NSEC3PARAM
NSEC3PARAM
.TP
.B PTR
Domain name pointer
.TP
.B RRSIG
Signature for a record set
.TP
.B SIG
Security signature
.TP
.B SOA
Marks the start of a zone of authority
.TP
.B SRV
Server selection
.TP
.B TA
DNSSEC trust authorities
.TP
.B TKEY
Transaction key
.TP
.B TLSA
TLSA
.TP
.B TSIG
Transaction signature
.TP
.B TXT
Text strings
.RE

.SH EXAMPLES

TBD

.SH FILES
.br
/etc/hosts
.br
/etc/resolv.conf

.SH SEE ALSO
.BR libgetdns (3),
.BR getdns_address (3),
.BR getdns_context (3),
.BR getdns_free_sync_request_memory (3), 
.BR getdns_hostname (3),
.BR getdns_service (3),



ChangeLog
* 2014-??-??: Version 0.1.6
  * Fix: linking against libev on FreeBSD
  * Fix: Let configure report problem on FreeBSD when configuring with
    libevent and libunbound <= 1.4.22 is not compiled with libevent.

* 2014-10-31: Version 0.1.5
  * Unit tests for transport settings
  * Fix: adhere to set maximum UDP payload size
  * API change: when no maximum UDP payload size is set, outgoing
    values will adhere to the suggestions in RFC 6891 and may follow
    a scheme that uses multiple values to maximize receptivity.
  * Stub mode use 1232 maximum UDP payload size when connecting to an
    IPv6 upstreams and 1432 with an IPv4 upstream.
  * Evaluate namespaces (or not) on a per query basis
  * GETDNS_NAMESPACE_LOCALNAMES namespace now gives just_address_answers
    only and does not mimic a DNS packet answer anymore
  * The add_opt_parameters extension
  * IPv6 scope_id support with link-local addresses.  Both with parsing
    /etc/resolv.conf and by providing them explicitly via
    getdns_context_set_upstream_recursive_servers
  * Query for A and AAAA simultaneously with return_both_v4_and_v6
  * GETDNS_TRANSPORT_TCP_ONLY_KEEP_CONNECTIONS_OPEN DNS transport
  * Fix: Answers without RRs in query secion (i.e. REFUSED)
  * Fix: Return empty response dict on timeout in async mode too
  * Move spec examples to spec subdirectory
  * Fix issue#76: Setting UDP Payload size below 512 should not error
  * Fix: Include OPT RR in response dict always (even without options)
  * TCP Fast open support (linux only).
    Enable with the --enable-tcp-fastopen configure option
  * Bump library version because of binary API change

* 2014-09-03: Version 0.1.4
  * Synchronous resolves now respect timeout setting,
  * On timeout *_sync functions now return GETDNS_RETURN_GOOD and a
    response dict with "status" GETDNS_RESPSTATUS_ALL_TIMEOUT>
  * Fix issue#50: getdns_dict_remove_name returns GETDNS_RETURN_GOOD on
    success.
  * Fix Issue#54: set_ub_dns_transport() not working
  * Fix Issue#49: Typo in documentation (thanks Stephane Bortzmeyer)
  * getdns_context_set_limit_outstanding_queries(),
    getdns_context_set_dnssec_allowed_skew() and
    getdns_context_set_edns_maximum_udp_payload_size() now working
  * <rr>_unknown rdata field for unknown or unsupported RR types
  * Temporarily disable timeout unit test 3 because of unpredictable results
  * Spec updated to version 0.507
  * Renamed "resolver_type" to "resolution_type" in dict returned from
    getdns_context_get_api_information()
  * Added GETDNS_RESPSTATUS_ALL_BOGUS_ANSWERS return code for with the
    dnssec_return_only_secure extension
  * Added support for CDS and CDNSKEY RR types, but needs ldns > 1.6.17 to
    be able to parse the wire format (not released yet at time of writing)
  * Added OPENPGPKEY RR type, but no rdata fields implementation yet
  * Updated spec to version 0.508 (September 2014)
  * Also chase NSEC and NSEC3 RRSIGs with dnssec_return_validation_chain

* 2014-06-25: Version 0.1.3
  * libtool chage, remove -release, added -version-info
  * Update specification to the June 2014 version (0.501)

* 2014-06-02: Version 0.1.2
  * Fixed rdata fields for MX
  * Expose only public API symbols
  * Updated manpages
  * specify_class extension
  * Build from separate build directory
  * Anticipate libunbound not returning the answer packet
  * Pretty print bindata's representing IP addresses
  * Anticipate absense of implicit DSO linking
  * Mention getdns specific options to configure in INSTALL
    Thanks Paul Hoffman
  * Mac OSX package built instructions for generic user in README.md
    Thanks Joel Purra
  * Fixed build problems on RHEL/CentOS due using libevent 1.x


* 2014-03-24 : Version 0.1.1
  * default to NOT build extensions (libev, libuv, libevent), handle
    --with/--without options to configure for them
  * Fixed some build/make nits
  * respect configure --docdir=X
  * Documentation/man page updates
  * Fix install and cpp guards in getdns_extra.h
  * Add method to switch between threads and fork mode for unbound
  * Fixes for libuv integration (saghul)
  * Fixes for calling getdns_destroy_context within a callback
  * Fixed signal related defines/decls


* 2014-02-25 : Version 0.1.0
  * Initial public release of the getdns API


src/common/getdns_general.c
/**
 *
 * /brief getdns core functions
 * 
 * This is the meat of the API
 * Originally taken from the getdns API description pseudo implementation.
 *
 */
/* The MIT License (MIT)
 * Copyright (c) 2013 Verisign, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * Much of this is based on / duplicated code from libevent evdns.  Credits to
 * Nick Mathewson and Niels Provos
 *
 * https://github.com/libevent/libevent/
 *
 * libevent dns is based on software by Adam Langly. Adam's original message:
 *
 * Async DNS Library
 * Adam Langley <agl@imperialviolet.org>
 * http://www.imperialviolet.org/eventdns.html
 * Public Domain code
 *
 * This software is Public Domain. To view a copy of the public domain dedication,
 * visit http://creativecommons.org/licenses/publicdomain/ or send a letter to
 * Creative Commons, 559 Nathan Abbott Way, Stanford, California 94305, USA.
 *
 * I ask and expect, but do not require, that all derivative works contain an
 * attribution similar to:
 *	Parts developed by Adam Langley <agl@imperialviolet.org>
 *
 * You may wish to replace the word "Parts" with something else depending on
 * the amount of original code.
 *
 * (Derivative works does not include programs which link against, run or include
 * the source verbatim in their source distributions)
 *
 * Version: 0.1b
 */

#include <getdns_context.h>
#include <ldns/ldns.h>

/* useful macros */
#define gd_malloc(sz) context->memory_allocator(sz)
#define gd_free(ptr) context->memory_deallocator(ptr)

struct getdns_dns_req;

typedef struct getdns_nameserver {
	evutil_socket_t socket;	 /* a connected UDP socket */
	struct sockaddr_storage address;
	ev_socklen_t addrlen;

	int failed_times;  /* number of times which we have given this server a chance */
	int timedout;  /* number of times in a row a request has timed out */
	struct event event;

	struct event timeout_event;  /* used to keep the timeout for */
    /* when we next probe this server. */
    /* Valid if state == 0 */
	/* Outstanding probe request for this nameserver, if any */
	struct getdns_dns_req *probe_request;
	char state;  /* zero if we think that this server is down */
	char choked;  /* true if we have an EAGAIN from this server's socket */
	char write_waiting;  /* true if we are waiting for EV_WRITE events */
	
    /* getdns context */
    getdns_context_t context;
    
	/* Number of currently inflight requests: used
	 * to track when we should add/del the event. */
	int requests_inflight;
} getdns_nameserver;

/* network request - state for a network request and referenced
 * by the the outbound_req
 */
typedef struct getdns_network_req {
	ldns_pkt *pkt;  /* the dns packet data */
	uint16_t request_type; /* query type */

	int reissue_count;
	int tx_count;  /* the number of times that this packet has been sent */
	
    /* not owned */
    struct nameserver *ns;	/* the server which we last sent it (unused) */
    getdns_dict *upstream_server;
    
	struct event timeout_event;
    
	getdns_transaction_t trans_id;  /* the transaction id */
	
	unsigned transmit_me :1;  /* needs to be transmitted */
    
	getdns_context_t context;
    
	struct getdns_dns_req *owner;

} getdns_network_req;

/* outbound request - manages recursion and stub reqs */
typedef struct getdns_dns_req {
    
	struct getdns_network_req *current_req;
    getdns_context_t context;
    
    uint16_t resolver_type;
    
    /* callback data */
    getdns_callback_t user_callback;
    void *user_pointer;
	
    
	int pending_cb; /* Waiting for its callback to be invoked; not
                     * owned by event base any more. */
    
	/* search not supported.. yet */

} getdns_dns_req;

/* stuff to make it compile pedantically */
#define UNUSED_PARAM(x) ((void)(x))

/* TODO: flags */
static ldns_pkt *create_new_pkt(getdns_context_t context,
                                const char* name,
                                uint16_t request_type,
                                struct getdns_dict* extensions) {
    ldns_pkt *pkt = NULL;
    ldns_rr_type type = (ldns_rr_type) request_type;
    ldns_pkt_query_new_frm_str(&pkt, name,
                                type,
                                LDNS_RR_CLASS_IN, 0);
    if (pkt) {
        /* id */
        ldns_pkt_set_id(pkt, ldns_get_random());
    }
    return pkt;
}

static void network_req_free(getdns_context_t context,
                             getdns_network_req* net_req) {
    if (!net_req) {
        return;
    }
    if (net_req->pkt) {
        ldns_pkt_free(net_req->pkt);
    }
    gd_free(net_req);
}

static getdns_network_req* network_req_new(getdns_context_t context,
                                           const char* name,
                                           uint16_t request_type,
                                           struct getdns_dict* extensions,
                                           getdns_transaction_t *transaction_id) {
    getdns_network_req *net_req = NULL;
    ldns_pkt *pkt = NULL;
    net_req = gd_malloc(sizeof(getdns_network_req));
    if (!net_req) {
        return NULL;
    }
    net_req->ns = NULL;
    net_req->pkt = NULL;
    net_req->context = context;
    net_req->request_type = request_type;
    
    /* create ldns packet */
    pkt = create_new_pkt(context, name, request_type, extensions);
    if (!pkt) {
        /* free up the req */
        network_req_free(context, net_req);
        return NULL;
    }
    net_req->pkt = pkt;
    net_req->trans_id = ldns_pkt_id(pkt);
    if (transaction_id) {
        *transaction_id = net_req->trans_id;
    }
    
    return net_req;
}

static void dns_req_free(getdns_context_t context,
                         getdns_dns_req* req) {
    if (!req) {
        return;
    }
    network_req_free(context, req->current_req);
    gd_free(req);
}

/* create a new dns req to be submitted */
static getdns_dns_req* dns_req_new(getdns_context_t context,
                                   const char* name,
                                   uint16_t request_type,
                                   struct getdns_dict *extensions,
                                   getdns_transaction_t *transaction_id) {
    getdns_dns_req *result = NULL;
    getdns_network_req *net_req = NULL;
    result = gd_malloc(sizeof(getdns_dns_req));
    if (result == NULL) {
        return NULL;
    }
    result->context = context;
    result->current_req = NULL;
    result->pending_cb = 0;
    result->resolver_type = context->resolution_type;
    
    /* create the initial network request */
    net_req = network_req_new(context, name, request_type,
                              extensions, transaction_id);
    if (!net_req) {
        dns_req_free(context, result);
        result = NULL;
    }
    
    result->current_req = net_req;
    net_req->owner = result;
    
    return result;
}

static getdns_return_t dict_to_sockaddr(getdns_dict* ns, struct sockaddr_storage* output) {
    struct getdns_bindata *address_type = NULL;
    struct getdns_bindata *address_data = NULL;
    uint16_t port = htons(53);
    memset(output, 0, sizeof(struct sockaddr_storage));
    output->ss_family = AF_UNSPEC;
    
    getdns_dict_get_bindata(ns, GETDNS_STR_ADDRESS_TYPE, &address_type);
    getdns_dict_get_bindata(ns, GETDNS_STR_ADDRESS_DATA, &address_data);
    if (!address_type || !address_data) {
        return GETDNS_RETURN_GENERIC_ERROR;
    }
    if (strcmp((char*) address_type->data, GETDNS_STR_IPV4)) {
        /* data is an in_addr_t */
        struct sockaddr_in* addr = (struct sockaddr_in*) output;
        addr->sin_family = AF_INET;
        addr->sin_port = port;
        memcpy(&(addr->sin_addr), address_data->data, address_data->size);
    } else {
        /* data is a v6 addr in host order */
        struct sockaddr_in6* addr = (struct sockaddr_in6*) output;
        addr->sin6_family = AF_INET6;
        addr->sin6_port = port;
        memcpy(&(addr->sin6_addr), address_data->data, address_data->size);
    }
    return GETDNS_RETURN_GOOD;
}

/* submit a new request to the event loop */
static getdns_return_t submit_new_dns_req(getdns_dns_req *request) {
    getdns_dict *nameserver = NULL;
    getdns_context_t context = request->context;
    struct sockaddr_storage sockdata;
    
    /* get first upstream server */
    getdns_list_get_dict(context->upstream_list, 0, &nameserver);
    if (!nameserver) {
        return GETDNS_RETURN_GENERIC_ERROR;
    }
    
    /* setup socket */
    if (dict_to_sockaddr(nameserver, &sockdata) != GETDNS_RETURN_GOOD) {
        return GETDNS_RETURN_GENERIC_ERROR;
    }
    evutil_socket_t sock = socket(sockdata.ss_family, SOCK_DGRAM, 0);
    evutil_make_socket_closeonexec(sock);
    evutil_make_socket_nonblocking(sock);
    
    return GETDNS_RETURN_GOOD;
}


/*
 * getdns_general
 */
getdns_return_t
getdns_general(
  getdns_context_t           context,
  const char                 *name,
  uint16_t                   request_type,
  struct getdns_dict         *extensions,
  void                       *userarg,
  getdns_transaction_t       *transaction_id,
  getdns_callback_t          callback
)
{
    /* Default to zero */
    if (transaction_id != NULL) {
        *transaction_id = 0;
    }
    if (!context || context->event_base == NULL ||
        callback == NULL ||
        context->resolution_type != GETDNS_CONTEXT_STUB) {
        /* Can't do async without an event loop
         * or callback
         *
         * Only supports stub right now.
         */
        return GETDNS_RETURN_BAD_CONTEXT;
    }
    

    /* create a req */
    getdns_dns_req *dns_req = dns_req_new(context, name, request_type,
                                          extensions, transaction_id);
    if (dns_req == NULL) {
        return GETDNS_RETURN_GENERIC_ERROR;
    }
    
    dns_req->user_callback = callback;
    dns_req->user_pointer = userarg;
    
    /* submit it */
    submit_new_dns_req(dns_req);

    UNUSED_PARAM(userarg);
    return GETDNS_RETURN_GOOD;
} /* getdns_general */


/*
 * getdns_address
 *
 */
getdns_return_t
getdns_address(
  getdns_context_t           context,
  const char                 *name,
  struct getdns_dict         *extensions,
  void                       *userarg,
  getdns_transaction_t       *transaction_id,
  getdns_callback_t          callback
)
{
    int cleanup_extensions = 0;
    if (!extensions) {
        extensions = getdns_dict_create();
        cleanup_extensions = 1;
    }
    getdns_dict_set_int(extensions,
                        GETDNS_STR_EXTENSION_RETURN_BOTH_V4_AND_V6,
                        GETDNS_EXTENSION_TRUE);

    getdns_return_t result = 
        getdns_general(context, name, GETDNS_RRTYPE_A,
                       extensions, userarg, transaction_id,
                       callback);
    if (cleanup_extensions) {
        getdns_dict_destroy(extensions);
    }
    return result;
} 

/* getdns_general.c */
